pkg/controllers/managementuser/snapshotbackpopulate/snapshotbackpopulate.go

```diff
@@ -77,58 +89,175 @@ func (h *handler) OnChange(key string, configMap *corev1.ConfigMap) (*corev1.Con
+				snapshot.SnapshotFile = cmGeneratedSnapshot.SnapshotFile
+				// restore the metadata and message as those may have been lost
+				snapshot.SnapshotFile.Metadata = md
+				snapshot.SnapshotFile.Message = msg
+				snapshot, err = h.etcdSnapshots.Update(snapshot)
+				if err != nil {
+					return configMap, fmt.Errorf("rkecluster %s/%s: error while updating etcd snapshot %s/%s: %w", cluster.Namespace, cluster.Name, snapshot.Namespace, snapshot.Name, err)
+				}
+				if snapshot.Status.Missing {
+					snapshot.Status.Missing = false
+					// the kube-apiserver only accepts status updates on deliberate subresource status updates which is why we have to double-call an update here if the missing is set incorrectly
+					if _, err := h.etcdSnapshots.UpdateStatus(snapshot); err != nil && !apierrors.IsNotFound(err) {
+						return configMap, fmt.Errorf("rkecluster %s/%s: error while setting status missing=false on etcd snapshot %s/%s: %w", cluster.Namespace, cluster.Name, snapshot.Namespace, snapshot.Name, err)
+					}
+				}
+			}
+		} else {
+			// create the snapshot in the mgmt cluster
+			logrus.Debugf("[snapshotbackpopulate] rkecluster %s/%s: creating etcd snapshot %s/%s as it differed from the actual snapshot config map", cluster.Namespace, cluster.Name, cmGeneratedSnapshot.Namespace, cmGeneratedSnapshot.Name)
+			logrus.Tracef("[snapshotbackpopulate] rkecluster %s/%s: creating etcd snapshot %s/%s: %+v", cluster.Namespace, cluster.Name, cmGeneratedSnapshot.Namespace, cmGeneratedSnapshot.Name, cmGeneratedSnapshot)
+			_, err = h.etcdSnapshots.Create(&cmGeneratedSnapshot)
+			if err != nil {
+				return configMap, fmt.Errorf("rkecluster %s/%s: error while creating etcd snapshot %s/%s: %w", cluster.Namespace, cluster.Name, cmGeneratedSnapshot.Namespace, cmGeneratedSnapshot.Name, err)
+			}
+		}
+	}
 	return configMap, nil
 }
 
-func (h *handler) configMapToSnapshots(configMap *corev1.ConfigMap) (result []rkev1.ETCDSnapshot, _ error) {
+func (h *handler) configMapToSnapshots(configMap *corev1.ConfigMap, cluster *provv1.Cluster) (result map[string]rkev1.ETCDSnapshot, _ error) {
+	clusterName := cluster.Name
+	clusterNamespace := cluster.Namespace
+	result = map[string]rkev1.ETCDSnapshot{}
 	for k, v := range configMap.Data {
 		file := &snapshotFile{}
 		if err := json.Unmarshal([]byte(v), file); err != nil {
 			logrus.Errorf("invalid non-json value in %s/%s for key %s in cluster %s", configMap.Namespace, configMap.Name, k, h.clusterName)
 			return nil, nil
 		}
+		snapshotName := clusterName + "-" + file.Name
+		// Validate that the corresponding machine for the node exists before creating the snapshot
 		snapshot := rkev1.ETCDSnapshot{
-			Name:      file.Name,
-			Location:  file.Location,
-			Metadata:  file.Metadata,
-			Message:   file.Message,
-			NodeName:  file.NodeName,
-			CreatedAt: file.CreatedAt,
-			Size:      file.Size,
-			Status:    file.Status,
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      snapshotName,
```

Comment:
Technically there is no point in setting `Name` here as we set `snapshot.Name` again at the end of loop's block. In fact we could just change `snapshot.Name = snapshotName` at line 258 to `snapshot.Name = clusterName + "-" + file.Name + fileSuffix`.